gcp-observability/src/main/java/io/grpc/gcp/observability/interceptors/LogHelper.java

```diff
@@ -331,67 +318,76 @@ GrpcLogRecord.Builder createTimestamp() {
-    T payload;
-    int size;
+  static final class PayloadBuilderHelper<T> {
+    T payloadBuilder;
     boolean truncated;
 
-    private PayloadBuilder(T payload, int size, boolean truncated) {
-      this.payload = payload;
-      this.size = size;
+    private PayloadBuilderHelper(T payload, boolean truncated) {
+      this.payloadBuilder = payload;
       this.truncated = truncated;
     }
   }
 
-  static PayloadBuilder<GrpcLogRecord.Metadata.Builder> createMetadataProto(Metadata metadata,
+  static PayloadBuilderHelper<Payload.Builder> createMetadataProto(Metadata metadata,
       int maxHeaderBytes) {
     checkNotNull(metadata, "metadata");
     checkArgument(maxHeaderBytes >= 0,
         "maxHeaderBytes must be non negative");
-    GrpcLogRecord.Metadata.Builder metadataBuilder = GrpcLogRecord.Metadata.newBuilder();
-    // This code is tightly coupled with io.grpc.observabilitylog.v1.GrpcLogRecord.Metadata
-    // implementation
-    byte[][] serialized = InternalMetadata.serialize(metadata);
+    Joiner joiner = Joiner.on(", ").skipNulls();
+    Payload.Builder payloadBuilder = Payload.newBuilder();
     boolean truncated = false;
     int totalMetadataBytes = 0;
-    if (serialized != null) {
-      // Calculate bytes for each GrpcLogRecord.Metadata.MetadataEntry
-      for (int i = 0; i < serialized.length; i += 2) {
-        String key = new String(serialized[i], Charsets.UTF_8);
-        byte[] value = serialized[i + 1];
-        if (NEVER_INCLUDED_METADATA.contains(key)) {
-          continue;
-        }
-        boolean forceInclude = ALWAYS_INCLUDED_METADATA.contains(key);
-        int metadataBytesAfterAdd = totalMetadataBytes + key.length() + value.length;
-        if (!forceInclude && metadataBytesAfterAdd > maxHeaderBytes) {
-          truncated = true;
-          continue;
-        }
-        metadataBuilder.addEntryBuilder()
-            .setKey(key)
-            .setValue(ByteString.copyFrom(value));
-        if (!forceInclude) {
-          // force included keys do not count towards the size limit
-          totalMetadataBytes = metadataBytesAfterAdd;
-        }
+    for (String key : metadata.keys()) {
+      if (NEVER_INCLUDED_METADATA.contains(key)) {
+        continue;
+      }
+      boolean forceInclude = ALWAYS_INCLUDED_METADATA.contains(key);
+      String metadataValue;
+      if (key.endsWith(Metadata.BINARY_HEADER_SUFFIX)) {
+        Iterable<byte[]> metadataValues =
+            metadata.getAll(Metadata.Key.of(key, Metadata.BINARY_BYTE_MARSHALLER));
+        List<String> numList = new ArrayList<String>();
+        metadataValues.forEach(
+            (element) -> {
+              numList.add(BASE64_ENCODING_OMIT_PADDING.encode(element));
+            });
+        metadataValue = joiner.join(numList);
+      } else {
+        Iterable<String> metadataValues = metadata.getAll(
+            Metadata.Key.of(key, Metadata.ASCII_STRING_MARSHALLER));
+        metadataValue = joiner.join(metadataValues);
```

Comment:
nit: I wonder if the common code can be moved out from the `if` and the `else` clause?

- the last `joiner.join` statement
- the first `metadata.getAll` 

May be create a util method for both ascii and bin cases?